src/main/scala/scalaz/nio/channels/SelectionKey.scala

```diff
@@ -0,0 +1,48 @@
+package scalaz.nio.channels
+
+import java.nio.channels.{SelectableChannel, SelectionKey => JSelectionKey}
+
+import scalaz.zio.IO
+
+class SelectionKey(private val selectionKey: JSelectionKey) {
+
+  def channel(): IO[Nothing, SelectableChannel] = // TODO: wrapper for SelectableChannel
+    IO.sync(selectionKey.channel())
+
+  def selector(): IO[Nothing, Selector] =
+    IO.sync(selectionKey.selector()).map(new Selector(_))
+
+  def isValid: IO[Nothing, Boolean] =
+    IO.sync(selectionKey.isValid)
+
+  def cancel(): IO[Exception, Unit] =
+    IO.syncException(selectionKey.cancel())
+
+  def interestOps(): IO[Exception, Int] =
+    IO.syncException(selectionKey.interestOps())
+
+  def interestOps(ops: Int): IO[Exception, SelectionKey] =
+    IO.syncException(selectionKey.interestOps(ops)).map(new SelectionKey(_))
+
+  def readyOps(): IO[Exception, Int] =
+    IO.syncException(selectionKey.readyOps())
+
+  def isReadable(): IO[Exception, Boolean] =
+    IO.syncException(selectionKey.isReadable)
+
+  def isWritable(): IO[Exception, Boolean] =
+    IO.syncException(selectionKey.isWritable)
+
+  def isConnectable(): IO[Exception, Boolean] =
+    IO.syncException(selectionKey.isConnectable)
+
+  def isAcceptable(): IO[Exception, Boolean] =
+    IO.syncException(selectionKey.isAcceptable)
+
+  def attach(ob: AnyRef): IO[Exception, AnyRef] =
```

Comment:
maybe `def attach[T](ob: T): IO[Exception, T] =`

Reply:
or `def attach(attachment: A): IO[Exception, A] =`, where `A` is passed as a type parameter to `SelectionKey`. This way it can be reused/bound to `attachment` method.

Reply:
@ysusuk actually `attach` takes a new attachment and returns an old one. I don't see an easy way to parametrize this without some big API changes.

Reply:
hmm, unless `attach` would do something like return a new `SelectionKey`...